# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/svg/SVGUtils.h
# Commit: a68110a8fddc
# Full Hash: a68110a8fddc7b6daa1ae5f2503e68c98a7cfd8a
# Author: Robert Longson <longsonr@gmail.com>
# Date: 2023-10-24 03:54:44
# Regressor Bug: 1765202
# File Overlap Count: 1
# Description:
#   Bug 1765202 part 2 - refactor MaskUsage so that its members are private r=emilio
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D191513
# ==============================================================================

diff -r 98bea2812b17 -r a68110a8fddc layout/svg/SVGUtils.h
--- a/layout/svg/SVGUtils.h	Mon Oct 23 20:11:10 2023 +0000
+++ b/layout/svg/SVGUtils.h	Mon Oct 23 20:11:11 2023 +0000
@@ -534,30 +534,61 @@
                                const gfxMatrix& aToCanvas,
                                const nsPresContext* presContext);
 
-  struct MaskUsage {
-    bool shouldGenerateMaskLayer;
-    bool shouldGenerateClipMaskLayer;
-    bool shouldApplyClipPath;
-    bool shouldApplyBasicShapeOrPath;
-    float opacity;
-
-    MaskUsage()
-        : shouldGenerateMaskLayer(false),
-          shouldGenerateClipMaskLayer(false),
-          shouldApplyClipPath(false),
-          shouldApplyBasicShapeOrPath(false),
-          opacity(0.0) {}
-
-    bool shouldDoSomething() {
-      return shouldGenerateMaskLayer || shouldGenerateClipMaskLayer ||
-             shouldApplyClipPath || shouldApplyBasicShapeOrPath ||
-             opacity != 1.0;
-    }
-  };
-
+  struct MaskUsage;
   static MaskUsage DetermineMaskUsage(const nsIFrame* aFrame,
                                       bool aHandleOpacity);
 
+  struct MOZ_STACK_CLASS MaskUsage {
+    friend MaskUsage SVGUtils::DetermineMaskUsage(const nsIFrame* aFrame,
+                                                  bool aHandleOpacity);
+
+    bool ShouldGenerateMaskLayer() const { return mShouldGenerateMaskLayer; }
+
+    bool ShouldGenerateClipMaskLayer() const {
+      return mShouldGenerateClipMaskLayer;
+    }
+
+    bool ShouldGenerateLayer() const {
+      return mShouldGenerateMaskLayer || mShouldGenerateClipMaskLayer;
+    }
+
+    bool ShouldGenerateMask() const {
+      return mShouldGenerateMaskLayer || mShouldGenerateClipMaskLayer ||
+             !IsOpaque();
+    }
+
+    bool ShouldApplyClipPath() const { return mShouldApplyClipPath; }
+
+    bool HasSVGClip() const {
+      return mShouldGenerateClipMaskLayer || mShouldApplyClipPath;
+    }
+
+    bool ShouldApplyBasicShapeOrPath() const {
+      return mShouldApplyBasicShapeOrPath;
+    }
+
+    bool IsOpaque() const { return mOpacity == 1.0f; }
+
+    bool IsTransparent() const { return mOpacity == 0.0f; }
+
+    float Opacity() const { return mOpacity; }
+
+    bool ShouldDoSomething() const {
+      return mShouldGenerateMaskLayer || mShouldGenerateClipMaskLayer ||
+             mShouldApplyClipPath || mShouldApplyBasicShapeOrPath ||
+             mOpacity != 1.0f;
+    }
+
+   private:
+    MaskUsage() = default;
+
+    float mOpacity = 0.0f;
+    bool mShouldGenerateMaskLayer = false;
+    bool mShouldGenerateClipMaskLayer = false;
+    bool mShouldApplyClipPath = false;
+    bool mShouldApplyBasicShapeOrPath = false;
+  };
+
   static float ComputeOpacity(const nsIFrame* aFrame, bool aHandleOpacity);
 
   /**
